src/DataTypes/DataTypeMap.cpp

```diff
@@ -0,0 +1,375 @@
+{
+    const auto & column_map = assert_cast<const ColumnMap &>(column);
+
+    const auto & nested_array = column_map.getNestedColumn();
+    const auto & nested_tuple = column_map.getNestedData();
+    const auto & offsets = nested_array.getOffsets();
+
+    size_t offset = offsets[row_num - 1];
+    size_t next_offset = offsets[row_num];
+
+    writeChar('{', ostr);
+    for (size_t i = offset; i < next_offset; ++i)
+    {
+        if (i != offset)
+            writeChar(',', ostr);
+        writer(key_type, nested_tuple.getColumn(0), i);
+        writeChar(':', ostr);
+        writer(value_type, nested_tuple.getColumn(1), i);
+    }
+    writeChar('}', ostr);
+}
+
+template <typename Reader>
+void DataTypeMap::deserializeTextImpl(IColumn & column, ReadBuffer & istr, bool need_safe_get_int_key, Reader && reader) const
+{
+    auto & column_map = assert_cast<ColumnMap &>(column);
+
+    auto & nested_array = column_map.getNestedColumn();
+    auto & nested_tuple = column_map.getNestedData();
+    auto & offsets = nested_array.getOffsets();
+
+    auto & key_column = nested_tuple.getColumn(0);
+    auto & value_column = nested_tuple.getColumn(1);
+
+    size_t size = 0;
+    assertChar('{', istr);
+
+    try
+    {
+        bool first = true;
+        while (!istr.eof() && *istr.position() != '}')
+        {
+            if (!first)
+            {
+                if (*istr.position() == ',')
+                    ++istr.position();
+                else
+                    throw Exception("Cannot read Map from text", ErrorCodes::CANNOT_READ_MAP_FROM_TEXT);
+            }
+
+            first = false;
+
+            skipWhitespaceIfAny(istr);
+
+            if (*istr.position() == '}')
+                break;
+
+            if (need_safe_get_int_key)
+            {
+                ReadBuffer::Position tmp = istr.position();
+                while (*tmp != ':' && *tmp != '}')
+                    ++tmp;
+                *tmp = ' ';
+                reader(key_type, key_column);
+            }
+            else
+            {
+                reader(key_type, key_column);
+                skipWhitespaceIfAny(istr);
+                assertChar(':', istr);
+            }
+
+            ++size;
+            skipWhitespaceIfAny(istr);
+            reader(value_type, value_column);
+
+            skipWhitespaceIfAny(istr);
+        }
+
+        offsets.push_back(offsets.back() + size);
+        assertChar('}', istr);
+    }
+    catch (...)
+    {
+        throw;
+    }
+}
+
+void DataTypeMap::serializeText(const IColumn & column, size_t row_num, WriteBuffer & ostr, const FormatSettings & settings) const
+{
+    serializeTextImpl(column, row_num, ostr,
+        [&](const DataTypePtr & nested_type, const IColumn & nested_column, size_t pos)
+        {
+            nested_type->serializeAsTextQuoted(nested_column, pos, ostr, settings);
```

Comment:
Ahh so this `nested_type` is not the `nested` type that is `Array(Tuple(key,  value))`, but the key or value type that is passed to the writer. This confused me for a moment...